src/fsharp/FSharp.Compiler.LanguageServer/Server.fs

```diff
@@ -12,17 +12,17 @@ type Server(sendingStream: Stream, receivingStream: Stream) =
     let converter = JsonOptionConverter() // special handler to convert between `Option<'T>` and `obj/null`.
     do formatter.JsonSerializer.Converters.Add(converter)
     let handler = new HeaderDelimitedMessageHandler(sendingStream, receivingStream, formatter)
-    let state = State()
-    let methods = Methods(state)
+    let methods = Methods()
     let rpc = new JsonRpc(handler, methods)
+    do methods.State.JsonRpc <- Some rpc
 
     member __.StartListening() =
         rpc.StartListening()
 
     member __.WaitForExitAsync() =
         async {
-            do! Async.AwaitEvent (state.Shutdown)
-            do! Async.AwaitEvent (state.Exit)
+            do! Async.AwaitEvent (methods.State.Shutdown)
```

Comment:
Do these wait forever, or is there a timeout.  If no timeout should there be one?

Reply:
They wait forever and they should (kind of) because the external process should stay alive as long as the client (e.g., VS) is up.

When the server is initialized it might be given a process ID.  Eventually we'll want to watch that process for exit and then exit ourselves; this would happen if the client exited without sending shutdown messages, like if the process was killed or it crashed.  Given that this is still experimental, it's a bug-level fix for later.